### PR TITLE
Avoid unnecessary scrolling when finding or highlighting all matches

### LINKED ISSUE #5561: Toggling the "Highlight all" search feature should not change the view position
When the _Highlight all_ search feature (checkbox in the [demo](http://mozilla.github.io/pdf.js/web/viewer.html), button in the extension) is toggled, then the view position seems to focus the first hit on the current page.
In my opinion, the toggle shouldn't move the view at all, especially if the user turns _Highlight all_ off. Changing the focus is also contrary to Firefox's built-in highlight behavior (I haven't tested other browsers).

(FF34, Windows 7, current Demo version: 1.0.1011)

### LINKED ISSUE #6746: Highlighting all matches causes the page to jump while scrolling
STR:
1 - open http://arxiv.org/pdf/1509.03893v1.pdf (tested only in Nigthly with e10s enabled, but I've had users report this to me in other setups);
2 - make sure "Page Fit" is selected as view, it's easier to reproduce like this but I've managed to reproduce it with other zoom/view factors applied;
3 - search (Ctrl+F) for "alice" and click the "Highlight all" button;
4 - press F3 three times so that again the first of three instances of "alice" is active (it does not occur when the first instance of "alice" is active for the first time!);
5 - click in the document to focus it and press "home" key to jump back to the beginning;
6 - scroll down with "page down" button until you are at around page 7 or 8 and it should jump ahead to page 10 where the first instance of "alice" appears.

I've found this is because [TextLayerBuilder.renderMatches scrolls the match into view indiscriminately](https://github.com/mozilla/pdf.js/blob/master/web/text_layer_builder.js#L213-L216) every time a page is (re)rendered, even while just scrolling (close) to it.

Here's how I've fixed it in my add-on FindBar Tweak: https://github.com/Quicksaver/FindBar-Tweak/commit/c48dc2d5efa480e163064d15e5ca3e3e2bd08c82
- PDFFindController.nextMatch sets a flag to show the current match;
- PDFFindController.updateMatchPosition only actually scrolls to the match when this flag is set, and resets it immediately after scrolling.

### PR DESCRIPTION
**Pull request by Quicksaver**
Read title, 'nuff said. :) Well ok, not really.

I left the commits separate for each different action they're actually doing:
- https://github.com/Quicksaver/pdf.js/commit/4876693b61dc8603bcdb7abb57c4954efd168095 - fixes #6746.
- ~~https://github.com/mozilla/pdf.js/commit/e1e2a6a3be3d869a04b11242e6ca84abb07ff2fa - is a bonus to the above commit, I don't believe there's any need to scroll on each match as it was doing. It was scrolling to the next line even if the next match was there, it can be disorienting to see the document move all the time like that. This is more in line with how Firefox and Chrome find in html pages as well; don't know about b2g honestly but I can't imagine it would be much different.~~
  - ~~I couldn't use the already existing getVisibleElements() method because I needed a way to distinguish between fully visible and only partially visible elements, and I'm not sure that could be done in there without some serious (and maybe ugly) hacking.~~
- https://github.com/Quicksaver/pdf.js/commit/35206c59e2763940cbb70910a0f4bac1dde475fc - fixes #5561, as suggested since it's in the neighborhood.

### CODE REVIEW

**Comment by Snuffleupagus** [comment]
Others could have another opinion, but I'm thinking that in order to speed up testing/reviewing of this PR, it might be advised to move the second commit (`Only scroll to match if it isn't already in view`) into a separate PR.

**Comment by Quicksaver** [comment]
I thought about doing that, but decided to leave it here since it fits into the whole "avoid extra scrolling on finding/highlights" theme of the branch. Either way, I ran out of coding time for today, I can only do that tomorrow (maybe wednesday) at the earliest if it ends up being necessary.

**Comment by Snuffleupagus** [comment]
A quick observation: It seems that toggling "Highlight all" now just disables the scrolling, however the active match still is still reset to the *first* one on the page.

**Comment by Quicksaver** [comment]
> A quick observation: It seems that toggling "Highlight all" now just disables the scrolling, however the active match still is still reset to the first one on the page.

True. Forgot about that. :( Just as a reference for when I get to it, I'm assuming it's best to squash that into the corresponding commit, instead of creating a new one?

**Comment by Snuffleupagus** [comment]
> Just as a reference for when I get to it, I'm assuming it's best to squash that into the corresponding commit, instead of creating a new one?

Yes, that is correct.

> I thought about doing that, but decided to leave it here since it fits into the whole "avoid extra scrolling on finding/highlights" theme of the branch.

Given that it changes the current behaviour quite a bit, and given that the viewer needs careful manual testing, removing the second commit will greatly improve the chances of this PR landing in a timely manner :)

File: web/pdf_find_controller.js [hunk 1]
```diff
@@ -265,6 +274,11 @@ var PDFFindController = (function PDFFindControllerClosure() {
         return;
       }
 
+      if(this.state.type !== 'findhighlightallchange') {
```
Comment (Snuffleupagus) [inline web/pdf_find_controller.js]: Nit: there should be a space between `if` and `(`.

File: web/pdf_find_controller.js [hunk 2]
```diff
@@ -330,14 +344,23 @@ var PDFFindController = (function PDFFindControllerClosure() {
      */
     updateMatchPosition: function PDFFindController_updateMatchPosition(
         pageIndex, index, elements, beginIdx, endIdx) {
+      // Only scroll matches into view when a find actually runs, and not while
+      // scrolling the document (rendering a page calls this as well).
+      if (!this.showCurrentMatch) { return; }
```
Comment (Snuffleupagus) [inline web/pdf_find_controller.js]: Nit: please avoid these kind of one-line `if` statements.

**Comment by Quicksaver** [comment]
Nits addressed, Highlight All no longer resets current match, removed second commit.

Note that I can't create a separate branch with the removed commit because its changes would conflict with those of this branch (so its branch would have to be based on this one, and it would contain these commits as well). If I understand things correctly, I need to either wait until this PR is pulled to create the new branch, or push the commit into this branch again.

BTW, I noticed that the initial commits in this PR were on mozilla/pdf.js, but now they're on my own repo... Did I do something bad? Does it even matter for pulling?

**Comment by Snuffleupagus** [comment]
I noticed that there's a slight change in behaviour with this PR, compared to the current master. Whether or not that's an issue I don't know, but it might a warrant discussion about the expected behaviour.
**Steps to reproduce:**
 1. Open the default `tracemonkey` pdf.
 2. Search for "trace".
 3. With the first search result highlighted, rotate the document *twice*.

Note that with the current master, the match is scrolled into view after the document is rotated, but not with this PR. (I noticed that e.g. Adobe Reader doesn't seem to do this either, so we need to decide what behaviour we want here.)

**Comment by Snuffleupagus** [comment]
It seems that when toggling "Highlight all", the `findResultsCount` keeps increasing.

**Comment by Quicksaver** [comment]
> It seems that when toggling "Highlight all", the findResultsCount keeps increasing.

Fixed. Also fixed a couple of other bugs I noticed while investigating this:
- toggling Highlight All before running a find would prevent it from clicking the next/previous button before changing the find query (forcing another new find).
- toggling Highlight All could keep the "pending" image in the findbar until actually finding again.

> Note that with the current master, the match is scrolled into view after the document is rotated, but not with this PR. (I noticed that e.g. Adobe Reader doesn't seem to do this either, so we need to decide what behaviour we want here.)

I believe this is another manifestation of #6746: rotating the document causes (parts of?) the page to render again, thus bringing the match into view again.

I'd vote to keep the new behavior as I think that's what's intended, "finding" is an action on its own and only that action should actually bring a match into view. For example, I don't want to open a doc, find something, then read it for half an hour, then do something completely unrelated like rotate it and see it scroll back to the match that I don't care about anymore.

**Comment by timvandermeij** [comment]
Closing since this is fixed in an alternative way.

**Pull request closed without merging**